rpc/internal/collector.service.go

```diff
@@ -96,3 +114,188 @@ func (s *collectorService) StreamEvents(srv PoplarEventCollector_StreamEventsSer
+type streamsCoordinator struct {
+	groups map[string]*streamGroup
+	mu     sync.Mutex
+}
+
+// streamGroup represents a group of streams writing to the same
+// ftdc/events.Collector. Each group is tracked by the streamCoordinator, which
+// only allows one streamGroup per collector. Stream groups coordinate writes
+// to the collector using a min heap that sorts based on the timestamp of each
+// event. The size of the min heap is never greater than the number of streams
+// in the group.
+type streamGroup struct {
+	collector        events.Collector
+	availableStreams []string
+	streams          map[string]chan error
+	eventHeap        *PerformanceHeap
+	mu               sync.Mutex
+}
+
+// addStream adds a new stream to the group for the given collector. If the
+// collector does not exist, an error is returned. If the stream group for the
+// collector does not exist, it is created.
+func (sc *streamsCoordinator) addStream(name string, registry *poplar.RecorderRegistry) error {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+
+	collector, ok := registry.GetEventsCollector(name)
+	if !ok {
+		return errors.New("collector '%s' not found")
+	}
+
+	group, ok := sc.groups[name]
+	if !ok {
+		group = &streamGroup{
+			collector:        collector,
+			availableStreams: []string{},
+			streams:          map[string]chan error{},
+			eventHeap:        &PerformanceHeap{},
+		}
+		heap.Init(group.eventHeap)
+		sc.groups[name] = group
+	}
+
+	group.mu.Lock()
+	defer group.mu.Unlock()
+
+	id := utility.RandomString()
+	group.streams[id] = make(chan error)
+	group.availableStreams = append(group.availableStreams, id)
+
+	return nil
+}
+
+// getStream returns a stream id and stream group for the given collector name.
+// If addStream was not called first, this will error.
+func (sc *streamsCoordinator) getStream(name string) (string, *streamGroup, error) {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+
+	group, ok := sc.groups[name]
+	if !ok {
+		return "", nil, errors.Errorf("no group for '%s'", name)
+	}
+
+	if len(group.availableStreams) == 0 {
+		return "", nil, errors.New("must register first")
+	}
+	id := group.availableStreams[0]
+	group.availableStreams = group.availableStreams[1:]
+
+	return id, group, nil
+}
+
+// addEvent writes the given event to the collector from the given stream. If
+// the stream does not exist an error is returned. Note that this function
+// blocks until all streams in the group have called addEvent and the minimum
+// timestamp can be guaranteed.
+func (sg *streamGroup) addEvent(id string, event *events.Performance) error {
+	sg.mu.Lock()
+	defer sg.mu.Unlock()
+
+	errChan, ok := sg.streams[id]
+	if !ok {
+		return errors.Errorf("stream %s does not exist in this stream group", id)
+	}
+
+	sg.eventHeap.SafePush(&performanceHeapItem{errChan: errChan, event: event})
+	if sg.eventHeap.Len() >= len(sg.streams) {
+		item := sg.eventHeap.SafePop()
+		go sg.sendError(item)
+	}
+
+	sg.mu.Unlock()
+	err := <-errChan
+	sg.mu.Lock()
```

Comment:
It's a bit weird to re-lock just so that the defer above unlocks properly. Would it be clearer to not defer the unlock since there's only two return statements within this function anyways?